bin/inx-chronicle/src/listener.rs

```diff
@@ -1,75 +1,198 @@
 
-#[derive(Debug)]
-pub enum InxEvent {
-    Message(inx::proto::Message),
-    LatestMilestone(inx::proto::Milestone),
+#[derive(Debug, Error)]
+pub enum InxListenerError {
+    #[error("The broker actor is not running")]
+    MissingBroker,
+    #[error(transparent)]
+    Inx(#[from] InxError),
+    #[error(transparent)]
+    Read(#[from] Status),
+    #[error(transparent)]
+    Runtime(#[from] RuntimeError),
 }
 
+#[derive(Debug)]
 pub struct InxListener {
-    // TODO: Consider storing `inx_client` to potentially restart some streams.
-    message_handle: JoinHandle<()>,
-    latest_milestone_handle: JoinHandle<()>,
+    config: InxConfig,
+    broker_handle: Addr<Broker>,
 }
 
 impl InxListener {
-    pub fn new(inx_client: InxClient<Channel>, broker_addr: BrokerAddr) -> Self {
-        let message_handle = {
-            let mut inx_client = inx_client.clone();
-            let broker_addr = broker_addr.clone();
-            tokio::spawn(async move { subscribe_messages(&mut inx_client, broker_addr).await })
-        };
-
-        let latest_milestone_handle = {
-            let mut inx_client = inx_client;
-            tokio::spawn(async move { subscribe_latest_milestone(&mut inx_client, broker_addr).await })
-        };
-
-        Self {
-            message_handle,
-            latest_milestone_handle,
+    pub fn new(config: InxConfig, broker_handle: Addr<Broker>) -> Self {
+        Self { config, broker_handle }
+    }
+}
+
+#[async_trait]
+impl Actor for InxListener {
+    type Data = InxClient<Channel>;
+    type Error = InxListenerError;
+
+    async fn init(&mut self, cx: &mut ActorContext<Self>) -> Result<Self::Data, Self::Error> {
+        info!("Connecting to INX...");
+        let mut inx_client = self.config.build().await?;
+
+        info!("Connected to INX.");
+        let response = inx_client.read_node_status(NoParams {}).await?;
+        info!("Node status: {:#?}", response.into_inner());
+
+        let message_stream = inx_client.listen_to_messages(MessageFilter {}).await?.into_inner();
+        let milestone_stream = inx_client.listen_to_latest_milestone(NoParams {}).await?.into_inner();
+
+        cx.spawn_actor_supervised::<InxStreamListener<inx::proto::Message>, _>(
+            InxStreamListener::new(self.broker_handle.clone())?.with_stream(message_stream),
+        )
+        .await;
+        cx.spawn_actor_supervised::<InxStreamListener<inx::proto::Milestone>, _>(
+            InxStreamListener::new(self.broker_handle.clone())?.with_stream(milestone_stream),
+        )
+        .await;
+        Ok(inx_client)
+    }
+}
+
+#[async_trait]
+impl HandleEvent<Report<InxStreamListener<inx::proto::Message>>> for InxListener {
+    async fn handle_event(
+        &mut self,
+        cx: &mut ActorContext<Self>,
+        event: Report<InxStreamListener<inx::proto::Message>>,
+        inx_client: &mut Self::Data,
+    ) -> Result<(), Self::Error> {
+        // TODO: Figure out why `cx.shutdown()` is not working.
```

Comment:
Is that figured out now?

Reply:
no :(

It is a problem because the function borrows the `Self` struct and carries that borrow over an await point, but I need to be able to specify that it should only need to borrow one field of the struct.